File: tools/dbench/answers/go_sonnet/s2/m2_cfg_chain.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

func main() {
	b, err := os.ReadFile("app.cfg")
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}
	var cfg map[string]interface{}
	if err := json.Unmarshal(b, &cfg); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	mode := "prod"
	if d, ok := cfg["debug"].(bool); ok && d {
		mode = "debug"
	}

	host := "localhost"
	if h, ok := cfg["host"].(string); ok && h != "" {
		host = h
	}

	port := 0
	if p, ok := cfg["port"].(float64); ok {
		port = int(p)
	}

	fmt.Printf("%s %s:%d\n", mode, host, port)
}

File: tools/dbench/answers/rust_haiku/s1/m2_cfg_chain.rs
use std::fs;

fn extract_json_bool(json: &str, key: &str) -> Option<bool> {
    let search = format!("\"{}\":", key);
    if let Some(pos) = json.find(&search) {
        let start = pos + search.len();
        let remainder = &json[start..].trim_start();
        if remainder.starts_with("true") {
            return Some(true);
        } else if remainder.starts_with("false") {
            return Some(false);
        }
    }
    None
}

fn extract_json_string(json: &str, key: &str) -> Option<String> {
    let search = format!("\"{}\":", key);
    if let Some(pos) = json.find(&search) {
        let start = pos + search.len();
        let remainder = &json[start..].trim_start();
        if remainder.starts_with('"') {
            let content = &remainder[1..];
            if let Some(end) = content.find('"') {
                return Some(content[..end].to_string());
            }
        }
    }
    None
}

fn main() {
    let content = fs::read_to_string("app.cfg").unwrap_or_default();

    let debug = extract_json_bool(&content, "debug").unwrap_or(false);
    let mode = if debug { "debug" } else { "prod" };

    let host = extract_json_string(&content, "host").unwrap_or_else(|| "localhost".to_string());

    let port = if let Some(port_str) = extract_json_string(&content, "port") {
        port_str
    } else {
        "8080".to_string()
    };

    println!("{} {}:{}", mode, host, port);
}
